### PR TITLE
feat: api for trace materialization

### LINKED ISSUE #6584: Add endpoint for attributes materialization for traces
Make the backend APIs available for use for materializing ad-hoc keys so that the query builder can use columns directly. This will also help us materialize without errors.

### PR DESCRIPTION
**Pull request by nityanandagohain**
feat: api for trace materialization

Fixes #6584 

Changes
* remane logfield to field.
* added `/api/v2/traces/fields` instead of `/api/v1/traces/fields` as it was colliding with search traces `/api/v1/traces/{traceId}`

----

> [!IMPORTANT]
> Introduces new API endpoints for trace materialization, renames `LogField` to `Field`, and updates validation and interfaces for enhanced trace handling.
> 
>   - **API Changes**:
>     - Added new endpoints `/api/v2/traces/fields` for GET and POST requests in `http_handler.go`.
>     - Implemented `traceFields()` and `updateTraceField()` in `APIHandler`.
>   - **Data Structure Changes**:
>     - Renamed `LogField` to `Field` in `response.go` and updated related functions.
>     - Updated `GetFieldsResponse` to use `Field` instead of `LogField`.
>   - **Validation**:
>     - Added `ValidateUpdateFieldPayloadV2()` in `validator.go` for new trace field validation.
>   - **Miscellaneous**:
>     - Updated `Reader` interface in `interface.go` to include `GetTraceFields()` and `UpdateTraceField()`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=SigNoz%2Fsignoz&utm_source=github&utm_medium=referral)<sup> for 2ca724a172ed965294f16d6d5b3f19d0e2c8e970. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 1]
```diff
@@ -2945,6 +2947,142 @@ func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.Upda
...
+		response.Selected = append(response.Selected, model.Field{
+			Name:     field.Key,
+			DataType: field.DataType.String(),
+			Type:     constants.Static,
+		})
+	}
+
+	// get attribute keys
+	attributes := []model.Field{}
+	query := fmt.Sprintf("SELECT DISTINCT tagKey, tagType, dataType from %s.%s group by tagKey, tagType, dataType", r.TraceDB, r.spanAttributesKeysTable)
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:2971]: DISTINCT is redundant if you use group by

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 2]
```diff
@@ -2945,6 +2947,142 @@ func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.Upda
...
+		return model.ForbiddenError(errors.New("removing a selected field is not allowed, please reach out to support."))
+	}
+
+	// name of the materialized column
+	colname := utils.GetClickhouseColumnNameV2(field.Type, field.DataType, field.Name)
+
+	// dataType and chDataType of the materialized column
+	// dataType: string => string, bool => bool, int64, float64 => number
+	// chDataType: string => String, bool => Bool, int64 => Float64, => float64 => Float64
+	chDataType := strings.ToUpper(string(field.DataType[0])) + field.DataType[1:]
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:3016]: Can you instead maintain a map? I think that would be cleaner.

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 3]
```diff
@@ -2945,6 +2947,142 @@ func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.Upda
...
+		chDataType = "Float64"
+	}
+
+	// typeName: tag => attributes, resource => resources
+	typeName := field.Type
+	if field.Type == string(v3.AttributeKeyTypeTag) {
+		typeName = constants.Attributes
+	} else if field.Type == string(v3.AttributeKeyTypeResource) {
+		typeName = constants.Resources
+	}
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:3029]: Maybe in another PR, we should add support for scope.

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 4]
```diff
@@ -2945,6 +2947,142 @@ func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.Upda
...
+		field.IndexGranularity = constants.DefaultLogSkipIndexGranularity
+	}
+	query := fmt.Sprintf("ALTER TABLE %s.%s ON CLUSTER %s ADD INDEX IF NOT EXISTS `%s_idx` (`%s`) TYPE %s  GRANULARITY %d",
+		r.TraceDB, r.traceLocalTableName,
+		r.cluster,
+		colname,
+		colname,
+		field.IndexType,
+		field.IndexGranularity,
+	)
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:3078]: Can you add minmax also for numbers?

File: pkg/query-service/app/clickhouseReader/reader.go (hunk 3 above)
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go:3029]: sure, not adding in this pr.

File: pkg/query-service/app/clickhouseReader/reader.go (hunk 2 above)
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go:3016]: Yeah, makes sense

File: pkg/query-service/app/clickhouseReader/reader.go (hunk 4 above)
Comment (nityanandagohain) [reply pkg/query-service/app/clickhouseReader/reader.go:3078]: added default minmax index for numbers

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 5]
```diff
@@ -2945,6 +2947,166 @@ func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.Upda
 	return nil
 }
 
+func (r *ClickHouseReader) GetTraceFields(ctx context.Context) (*model.GetFieldsResponse, *model.ApiError) {
+	// response will contain top level fields from the otel log model
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:2951]:
```suggestion
	// response will contain top level fields from the trace model
```

File: pkg/query-service/app/clickhouseReader/reader.go [hunk 6]
```diff
@@ -2945,6 +2947,166 @@ func (r *ClickHouseReader) UpdateLogField(ctx context.Context, field *model.Upda
...
+	}
+
+	// add a default minmax index for numbers
+	if dataType == "number" {
+		query = fmt.Sprintf("ALTER TABLE %s.%s ON CLUSTER %s ADD INDEX IF NOT EXISTS `%s_minmax_idx` (`%s`) TYPE minmax  GRANULARITY %d",
+			r.TraceDB, r.traceLocalTableName,
+			r.cluster,
+			colname,
+			colname,
+			field.IndexGranularity,
```
Comment (srikanthccv) [inline pkg/query-service/app/clickhouseReader/reader.go:3099]: Can you change granularity to 1 for minmax, it requires the least amount of resources so it should be fine to set it to 1.

**Review by srikanthccv — COMMENTED** [review]
minor comments

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**